onnxruntime/core/providers/nnapi/nnapi_builtin/model.cc

```diff
@@ -0,0 +1,169 @@
+
+Model::~Model() {
+  if (execution_)
+    nnapi_->ANeuralNetworksExecution_free(execution_);
+
+  nnapi_->ANeuralNetworksCompilation_free(compilation_);
+  nnapi_->ANeuralNetworksModel_free(model_);
+}
+
+void Model::AddInput(const std::string& name, const android::nn::wrapper::OperandType& operand_type) {
+  input_names_.push_back(name);
+  operand_types_.insert({name, operand_type});
+}
+
+void Model::AddOutput(const std::string& name, const android::nn::wrapper::OperandType& operand_type) {
+  output_names_.push_back(name);
+  operand_types_.insert({name, operand_type});
+}
+
+const std::vector<std::string>& Model::GetInputs() const {
+  return input_names_;
+}
+
+const std::vector<std::string>& Model::GetOutputs() const {
+  return output_names_;
+}
+
+const android::nn::wrapper::OperandType& Model::GetInputType(const std::string& name) const {
+  return operand_types_.at(name);
+}
+
+const android::nn::wrapper::OperandType Model::GetOutputType(const std::string& name) const {
+  const auto& output_type = operand_types_.at(name);
+  android::nn::wrapper::OperandType type(
+      output_type.type, shaper_for_exeuction_[name], output_type.operandType.scale, output_type.operandType.zeroPoint);
+
+  return type;
+}
+
+void Model::SetInputMap(std::unordered_map<std::string, size_t>&& input_map) {
+  input_map_ = std::move(input_map);
+}
+
+void Model::SetOutputMap(std::unordered_map<std::string, size_t>&& output_map) {
+  output_map_ = std::move(output_map);
+}
+
+size_t Model::GetMappedInputIdx(const std::string& name) const {
+  return input_map_.at(name);
+}
+
+size_t Model::GetMappedOutputIdx(const std::string& name) const {
+  return output_map_.at(name);
+}
+
+void Model::SetInputBuffer(const int32_t index, const InputOutputBuffer& input) {
+  if (!prepared_for_exe_) PrepareForExecution();
+  THROW_ON_ERROR(nnapi_->ANeuralNetworksExecution_setInput(
+      execution_, index, &input.type.operandType, input.buffer, input.type.GetOperandBlobByteSize()));
+}
+
+void Model::SetOutputBuffer(const int32_t index, const InputOutputBuffer& output) {
+  if (!prepared_for_exe_) PrepareForExecution();
+  THROW_ON_ERROR(nnapi_->ANeuralNetworksExecution_setOutput(
+      execution_, index, &output.type.operandType, output.buffer, output.type.GetOperandBlobByteSize()));
+}
+
+void Model::PrepareForExecution() {
+  ORT_ENFORCE(nullptr != compilation_,
+              "Error in PrepareForExecution, compilation_ is null");
+
+  // Copy the shaper for calculate the dynamic output shape
+  // based on the input shape
+  shaper_for_exeuction_ = shaper_;
+
+  THROW_ON_ERROR(
+      nnapi_->ANeuralNetworksExecution_create(compilation_, &execution_));
+  prepared_for_exe_ = true;
+}
+
+void Model::ResetExecution() {
+  nnapi_->ANeuralNetworksExecution_free(execution_);
+  execution_ = nullptr;
+  shaper_for_exeuction_.Clear();
+  prepared_for_exe_ = false;
+}
+
+void Model::Predict() {
+  if (!prepared_for_exe_) PrepareForExecution();
+
```

Comment:
Wondering how this works if there are concurrent calls to InferenceSession::Run given there's only one instance of the EP. Is any locking needed to handle that correctly?

Reply:
We currently do not handle concurrent calls to the same model object, 
Will have a TODO to lock the model until the predict is run

Reply:
Would be good to add a comment in the code so that is captured (ideally linking to a Task for the actual work)

---
In reply to: [444561155](https://github.com/microsoft/onnxruntime/pull/4287#discussion_r444561155) [](ancestors = 444561155)

Reply:
Added a comment to the compute_func, since the use of model itself should be blocked to a single thread, not only Predict() since you can setinput from other thread if you can call the concurrently


---
In reply to: [444751769](https://github.com/microsoft/onnxruntime/pull/4287#discussion_r444751769) [](ancestors = 444751769,444561155)